106 ensure node version

evanplaice opened the pull request:
This PR closes #106

#### What does this PR do?

- adds semver as a devDependency
- adds /src/util/checkVersion.js
- adds a 'check-version' npm script
- runs a version check against the engine version specified in `package.json`
  - before `npm install`
  - before `npm start`

#### How does this PR make you feel?

![](https://media.giphy.com/media/S24vAZeES2INy/giphy.gif)

evanplaice commented:
CI doesn't want to install semver. Going to implement the version check without it.

evanplaice commented:
Netlify CI fails because the node version specified in the config needs to be >= 8.9.0.

8.9.3 is the latest LTS release.

wdoug commented on .travis.yml:
```diff
@@ -1,6 +1,6 @@
 language: node_js
 node_js:
-  - 8.4.0
+  - 8.9.3
```
Let's update the `.nvmrc` file as well so people using nvm will have the correct version.

wdoug commented on package.json:
```diff
@@ -28,13 +27,15 @@
     "gc": "cd ./server && dotenv -e ../.env.local graphcool",
-    "deployProdBackend": "cd ./server && graphcool deploy"
+    "deployProdBackend": "cd ./server && graphcool deploy",
+    "check-version": "node ./src/utils/checkVersion.js"
```
I would probably put this in the `bin` directory as a script that is associated with setup, whereas the way this repo is set up the `src` directory is where the client code lives.

evanplaice replied:
Good point, not sure how I missed /bin before. Isn't the usual convention to put CLI scripts in /.bin?

wdoug commented:
I just came across this [check-node-version](https://www.npmjs.com/package/check-node-version) package. Would that simplify things for us?

evanplaice commented:
I originally used semver to do the check. For some reason CI wouldn't install the package. That package has 7 top-level deps incl semver. IMO, it's better to avoid piling packages onto the dep tree for such a simple use case.

The current implementation will ensure greater-than or greater-than-or-equal. Anything else and it'll throw an error.

evanplaice commented:
ESLint is complaining about the console statements. Is there any special way I should log messages to the console, or should I just add an exception to this script?

wdoug commented:
Holy moly, 7 dependencies? In that case, I'm cool to keep this approach.

I would just disable the `no-console` rule for the bin directory. 🤔 ...or maybe we should just disable that from the whole repo... either seems fine to me.

evanplaice commented:
The no-console rule makes sense for client code. You wouldn't want a forgotten debug statement spitting output to the console in the background. CLI scripts are the obvious exception because that's the only way to provide feedback to the user.

wdoug commented on bin/checkVersion.js:
```diff
@@ -0,0 +1,48 @@
+/*eslint no-console: "off"*/
```
I think we could set this for the whole `bin` directory (https://github.com/codefordenver/members/blob/master/bin/.eslintrc), but this is fine for now.

wdoug reviewed (approved):
:shipit:

wdoug commented:
@evanplaice you may squash and merge when ready

evanplaice committed:
Ensure proper npm version before `npm install` and `npm start`


Add the checkVersion.js util script
Update npm scripts to use check-version
Upgrade package.node.engine version
Upgrade Travis-CI node version
Upgrade .nvmrc node version

evanplaice commented:
OK, squashed and ready to go!
